src/bare-metal/microcontrollers/examples/src/bin/hal.rs

```diff
@@ -0,0 +1,44 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// ANCHOR: Example
+#![no_main]
+#![no_std]
+
+extern crate panic_halt as _;
+
+use cortex_m_rt::entry;
+use nrf52833_hal::{
```

Comment:
You could save a bit of vertical space by importing the items over multiple lines like this:

```rust
use nrf52833_hal::gpio::{p0, Level};
use nrf52833_hal::pac::Peripherals;
use nrf52833_hal::prelude::*;
```

Reply:
I guess, but it fits on screen fine as it is and is easier to read.

Reply:
I have consistently avoided nested imports in the rest of the class and I would like you to follow the same style in the new code. I do that because I personally find such imports very annoying to read (they require me to build kind of mental stack when reading vs having a flat list of imports with helpful repetition).

As for fitting on the screen fine, I see it take up the full vertical height after you scroll it into view. So yeah, it fits, but only barely :smile: 

![image](https://user-images.githubusercontent.com/89623/224476357-14525765-32a9-48b9-83d7-3c0f002b4daf.png)

Without scrolling, this is what I see:

![image](https://user-images.githubusercontent.com/89623/224476538-6a79f64c-59d9-4c0f-8c8d-0532a43578a9.png)

I would suggest removing things like `// GPIO port 0.` and the blank line above it.

Reply:
Do you have a formatter rule for your preferred style? It seems like you're nesting some imports but not others. I prefer to nest as much as possible to save space and make it easier to read, but I'm happy to consistently not nest at all. Going half way and having some things nested but not others seems inconsistent and annoying.

FWIW this page fits on screen fine for me without scrolling, but some of the earlier ones (e.g. the Raw MMIO and PAC examples) don't. I don't think there's any good way around that, some of these examples just are quite long, even being as minimal as possible. I'll see how it goes when I actually teach the course but I don't think having to scroll a bit is going to be a big deal.

Reply:
> Do you have a formatter rule for your preferred style?

I would like to think that I use [`import_granularity: "Module"`](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=modul#Module%5C%3A) for the code snippets. Within that style, I prefer to import modules over individual items when the module name is short. I know it's not a hard rule :smile: 

> FWIW this page fits on screen fine for me without scrolling

So, I suppose the aspect ratio helper is affected by zoom levels...? I guess I should have know that :smile: 

> I'll see how it goes when I actually teach the course but I don't think having to scroll a bit is going to be a big deal.

Yeah, it's near impossible to keep everything short when you want to show real world examples. Scrolling being a problem was given to me as direct feedback in Munich and that's why I'm trying to become better at avoiding it.